hnn_core/lfp.py

```diff
@@ -0,0 +1,268 @@
+        H = np.dot(b, a) / norm_a  # NB can be negative
+        L = H + norm_a
+        R2 = np.dot(b, b) - H ** 2  # NB squares
+        # To avoid numerical errors when electrode is placed (anywhere) on the
+        # section axis, enforce minimal axial distance
+        R2 = max(R2, (section.diam / 2.0 + 0.1) ** 2)
+
+        if L < 0 and H < 0:  # electrode is "behind" section
+            num = np.sqrt(H ** 2 + R2) - H  # == norm(b) - H
+            denom = np.sqrt(L ** 2 + R2) - L
+        elif L > 0 and H < 0:  # electrode is "on top of" section
+            num = (np.sqrt(H ** 2 + R2) - H) * (L + np.sqrt(L ** 2 + R2))
+            denom = R2
+        else:  # electrode is "ahead of" section
+            num = np.sqrt(L ** 2 + R2) + L
+            denom = np.sqrt(H ** 2 + R2) + H  # == norm(b) + H
+
+        phi = np.log(num / denom) / norm_a
+
+    # [dis]: um; [sigma]: S / m
+    # [phi / sigma] = [1/dis] / [sigma] = 1 / [dis] x [sigma]
+    # [dis] x [sigma] = um x (S / m) = 1e-6 S
+    # transmembrane current returned by _ref_i_membrane_ is in [nA]
+    # ==> 1e-9 A x (1 / 1e-6 S) = 1e-3 V = mV
+    # ===> multiply by 1e3 to get uV
+    return 1000.0 * phi / (4.0 * np.pi * sigma)
+
+
+class _TransmembraneCurrentHandler:
+    """Transmembrane net current handler class.
+
+    The handler is set up to maintain a vector of membrane currents at at every
+    inner segment of every section of every cell on each CVODE integration
+    step. In addition, it records a time vector of sample times. This class
+    must be instantiated and attached to the network during the building
+    process. It is used in conjunction with the calculation of extracellular
+    potentials.
+
+    Parameters
+    ----------
+    pc : instance of h.ParallelContext
+        Used to keep track of time.
+    cvode : instance of h.CVode
+        Multi order variable time step integration method.
+    """
+    def __init__(self, pc=None, cvode=None):
+
+        self.pc = pc
+        if self.pc is None:
+            self.pc = h.ParallelContext()
+
+        # ordered list of h.Sections on this rank (if running in parallel)
+        self.secs_on_rank = _get_sections_on_this_rank()
+        # np.array of number of segments for each section, ordered as above
+        self.segment_counts = np.array(_get_segment_counts(self.secs_on_rank))
+
+        # pointers assigned to _ref_i_membrane_ at each EACH internal segment
+        self.imem_ptrvec = h.PtrVector(self.segment_counts.sum())
+        # placeholder into which pointer values are read on each sim step
+        imem_vec_len = int(self.imem_ptrvec.size())
+        self.imem_vec = h.Vector(imem_vec_len)
+
+        count = 0
+        for sec in self.secs_on_rank:
+            for seg in sec:  # section end points (0, 1) not included
+                # set Nth pointer to the net membrane current at this segment
+                self.imem_ptrvec.pset(count, sec(seg.x)._ref_i_membrane_)
```

Comment:
are you sure everything is correct here? Pointers are often to blame for segmentation faults.

Reply:
For a single trial, this works and produces correct results. So the pointers are correct on initialisation, I'm sure.